junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherDiscoveryRequestBuilder.java

```diff
@@ -268,14 +273,24 @@ private LauncherDiscoveryListener getLauncherDiscoveryListener(ConfigurationPara
 				.map(value -> LauncherDiscoveryListeners.fromConfigurationParameter(
 					DEFAULT_DISCOVERY_LISTENER_CONFIGURATION_PROPERTY_NAME, value)) //
 				.orElseGet(LauncherDiscoveryListeners::abortOnFailure);
-		if (discoveryListeners.isEmpty()) {
+
+		List<LauncherDiscoveryListener> serviceLoaderListeners = configurationParameters.getBoolean(
+			DISCOVERY_LISTENERS_AUTODETECTION_ENABLED_PROPERTY_NAME).filter(p -> p).map(
+				p -> StreamSupport.stream(
+					new ServiceLoaderLauncherDiscoveryListenerRegistry().loadListeners().spliterator(), false).collect(
+						Collectors.<LauncherDiscoveryListener> toList())).orElse(Collections.emptyList());
+
+		if (discoveryListeners.isEmpty() && serviceLoaderListeners.isEmpty()) {
 			return defaultDiscoveryListener;
 		}
-		if (discoveryListeners.contains(defaultDiscoveryListener)) {
+		if (discoveryListeners.contains(defaultDiscoveryListener) && serviceLoaderListeners.isEmpty()) {
 			return LauncherDiscoveryListeners.composite(discoveryListeners);
 		}
 		List<LauncherDiscoveryListener> allDiscoveryListeners = new ArrayList<>(discoveryListeners);
-		allDiscoveryListeners.add(defaultDiscoveryListener);
+		allDiscoveryListeners.addAll(serviceLoaderListeners);
+		if (!allDiscoveryListeners.contains(defaultDiscoveryListener)) {
+			allDiscoveryListeners.add(defaultDiscoveryListener);
+		}
 		return LauncherDiscoveryListeners.composite(allDiscoveryListeners);
```

Comment:
I don't think this optimization is necessary. Let's get rid of the first `if`.

Reply:
Sure no problem, I was not entirely sure about this part as well, it didn't feel quite right anyways. I'll make a small optimisation so that the list is created with a correct size at least (to avoid resizes)